presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java

```diff
@@ -0,0 +1,288 @@
+
+            Set<Set<RowExpression>> combinations = allInequalities.size() < 3 ?
+                    ImmutableSet.of(allInequalities) :
+                    Sets.combinations(allInequalities, 2);
+            for (Set<RowExpression> pair : combinations) {
+                Iterator<RowExpression> it = pair.iterator();
+                compareAndExtractInequalities(it.next(), it.next()).ifPresent(inequalitiesInferredInCurrentTraversal::add);
+            }
+        }
+        while (!inferredInequalityExpressions.containsAll(inequalitiesInferredInCurrentTraversal));
+
+        return ImmutableSet.copyOf(inferredInequalityExpressions);
+    }
+
+    private void validateConjunct(RowExpression expression)
+    {
+        checkArgument(expression instanceof CallExpression, "conjunct has to be a CallExpression");
+        CallExpression conjunct = (CallExpression) expression;
+        checkArgument(conjunct.getArguments().size() == 2, "callExpressionConjunct has to have exactly 2 arguments");
+
+        Optional<OperatorType> optOperatorType = functionAndTypeManager.getFunctionMetadata(conjunct.getFunctionHandle()).getOperatorType();
+        checkArgument(optOperatorType.isPresent() && optOperatorType.get().isComparisonOperator(), "optOperatorType has to be present and has to be a comparison operator");
+
+        OperatorType operatorType = optOperatorType.get();
+        checkArgument(operatorType.equals(LESS_THAN) || operatorType.equals(LESS_THAN_OR_EQUAL), "operatorType has to be < or <=");
+    }
+
+    // If the join is an outer join, we need to make sure any inferred predicate is on the inner side only
+    // (i.e. does not reference any 'outerVariables').
+    //private Optional<RowExpression> compareAndExtractInequalities(int firstConjunctIndex, int secondConjunctIndex, ImmutableList<RowExpression> inequalities)
+    private Optional<RowExpression> compareAndExtractInequalities(RowExpression expression1, RowExpression expression2)
+    {
+        CallExpression firstConjunct = (CallExpression) expression1;
+        OperatorType firstOperatorType = functionAndTypeManager.getFunctionMetadata(firstConjunct.getFunctionHandle()).getOperatorType().get();
+
+        CallExpression secondConjunct = (CallExpression) expression2;
+        OperatorType secondOperatorType = functionAndTypeManager.getFunctionMetadata(secondConjunct.getFunctionHandle()).getOperatorType().get();
+
+        // Get the inference operator, or empty if no inference possible.
+        Optional<OperatorType> inferenceOperatorType = getComparisonInferenceOperatorType(secondOperatorType, firstOperatorType);
+        if (!inferenceOperatorType.isPresent()) {
+            return Optional.empty();
+        }
+
+        Optional<RowExpression> inferredFirstArgument = Optional.empty();
+        Optional<RowExpression> inferredSecondArgument = Optional.empty();
+        Set<VariableReferenceExpression> variablesReferencedInInferredPredicate = new HashSet<>();
```

Comment:
That would lead to a "variable not initialized" error when we use it later. But with the suggested rewrite, I think the code is easier to parse, and its usage is guarded by "inferredFirstArgument.isPresent()". Wdyt?